feat(mobile): ios widgets

bwees opened the pull request:
## Description

Adds iOS Widgets for random (whole library or from album) and memories (falling back to random if no memories available)

## How Has This Been Tested?

Tested on iOS 18.5

<img width="414" alt="image" src="https://github.com/user-attachments/assets/653947bf-3e4e-4290-815a-1acbb20af188" />


## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if applicable
- [x] I have no unrelated changes in the PR.
- [x] I have confirmed that any new dependencies are strictly necessary.
- [x] I have written tests for new code (if applicable)
- [x] I have followed naming conventions/patterns in the surrounding code
- [x] All code in `src/services/` uses repositories implementations for database calls, filesystem operations, etc.
- [x] All code in `src/repositories/` is pretty basic/simple and does not have any immich specific logic (that belongs in `src/services/`)

bwees committed:
feat: working widgets

bwees committed:
chore/feat: cleaned up API, added album picker to random widget

bwees committed:
album filtering for requests

bwees committed:
check album and throw if not found

bwees committed:
fix app IDs and project configuration

bwees committed:
switch to repository/service model for updating widgets

bwees committed:
fix: remove home widget import

bwees committed:
revert info.plist formatting changes

bwees committed:
ran swift-format on widget code

bwees committed:
more formatting changes (this time run from xcode)

bwees committed:
show memory on widget picker snapshot

shenlong-tanwen commented on mobile/lib/repositories/widget.repository.dart:
```diff
@@ -0,0 +1,24 @@
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:immich_mobile/interfaces/widget.interface.dart';
+
+final widgetRepositoryProvider = Provider((ref) => WidgetRepository());
```
```suggestion
final widgetRepositoryProvider = Provider((_) => WidgetRepository());
```

shenlong-tanwen commented on mobile/lib/repositories/widget.repository.dart:
```diff
@@ -0,0 +1,24 @@
+  WidgetRepository();
+
+  @override
+  void saveWidgetData(String key, String value) {
```
Since the methods are in the `WidgetRepository`, should we drop "Widget" from the method names? So, instead of `saveWidgetData`, it'll just be `saveData`

shenlong-tanwen commented on mobile/lib/services/widget.service.dart:
```diff
@@ -0,0 +1,34 @@
+    _repository.saveWidgetData(kWidgetAuthToken, sessionKey);
+  }
+
+  void clearWidgetData() {
```
Can we call this `clearCredentials` instead?

shenlong-tanwen commented on mobile/lib/providers/auth.provider.dart:
```diff
@@ -112,6 +124,18 @@ class AuthNotifier extends StateNotifier<AuthState> {
+
+    // wait 3 seconds to ensure the widget is updated, dont block
+    Future.delayed(const Duration(seconds: 3), () {
+      _widgetService.updateWidget('com.immich.widget.random');
+      _widgetService.updateWidget('com.immich.widget.memory');
```
If we are going to update all widgets after a change in the credential, should we make it the responsibility of the service to reload the widgets rather than making the consumer of it? Also, the widget IDs can be constants

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/MemoryWidget.swift:
```diff
@@ -0,0 +1,152 @@
+    do {
+      let memories = try await api.fetchMemory(for: Date.now)
+
+      for memory in memories {
+        for asset in memory.assets {
+          if asset.type == "IMAGE" {
+            entries.append(
+              try await buildEntry(
+                api: api,
+                asset: asset,
+                hourOffset: entries.count,
+                subtitle: getYearDifferenceSubtitle(assetYear: memory.data.year)
```
This could be a task group

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/MemoryWidget.swift:
```diff
@@ -0,0 +1,152 @@
+      print("Failed to fetch from server: \(error)")
+    }
+
+    // If we didnt add any memory images, default to 12 hours of random photos
+    if entries.count == 0 {
+      entries.append(
+        contentsOf: (try? await generateRandomEntries(
+          api: api,
+          now: now,
+          count: 12
+        )) ?? []
+      )
+    }
```
This one can be a task group as well

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/MemoryWidget.swift:
```diff
@@ -0,0 +1,152 @@
+      provider: ImmichMemoryProvider()
+    ) { entry in
+      ImmichWidgetView(entry: entry)
+        .containerBackground(.black, for: .widget)
```
`.black` does not work well in light mode, can we use `.regularMaterial` or something dynamic?

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/RandomWidget.swift:
```diff
@@ -0,0 +1,152 @@
+    "Choose an album to show images from"
+  }
+
+  // An example configurable parameter.
```
```suggestion
```

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/RandomWidget.swift:
```diff
@@ -0,0 +1,152 @@
+      return ImageEntry(date: Date(), image: nil, error: .fetchFailed)
+    }
+
+    guard let image = try? await api.fetchImage(asset: demoImage) else {
+      return ImageEntry(date: Date(), image: nil, error: .fetchFailed)
+    }
```
I believe we can use `buildEntry` here

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/RandomWidget.swift:
```diff
@@ -0,0 +1,152 @@
+      configuration.album?.id != "NONE" ? configuration.album?.id : nil
+    if albumId != nil {
+      // make sure the album exists on server, otherwise show error
+      guard let albums = try? await api.fetchAlbums() else {
+        entries.append(ImageEntry(date: now, image: nil, error: .fetchFailed))
+        return Timeline(entries: entries, policy: .atEnd)
+      }
+
+      if !albums.contains(where: { $0.id == albumId }) {
+        entries.append(ImageEntry(date: now, image: nil, error: .albumNotFound))
+        return Timeline(entries: entries, policy: .atEnd)
+      }
+    }
```
Is the reason for making a call to the server to fetch the albums and not using the local cache is to prevent any stale data? If not, can't we use the cache here?

bwees replied:
we are avoiding as much swift <-> dart interop until the sqlite migration is complete

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/RandomWidget.swift:
```diff
@@ -0,0 +1,152 @@
+      }
+    }
+
+    entries.append(
+      contentsOf: (try? await generateRandomEntries(
+        api: api,
+        now: now,
+        count: 24,
+        albumId: albumId
+      ))
+        ?? []
+    )
```
Task group :)

bwees commented on mobile/ios/WidgetExtension/widgets/MemoryWidget.swift:
```diff
@@ -0,0 +1,152 @@
+  // TODO: make no configuration
+}
+
+struct ImmichMemoryProvider: AppIntentTimelineProvider {
```
AppIntentTimelineProvider is the only one that supports async methods for timelines and snapshots. The code gets way more messy if I use TimelineProvide because then everything has to be a closure-based instead of the nice async await code.

shenlong-tanwen replied:
Ah that makes sense, Can you maybe add a comment as to why we use this then?

shenlong-tanwen commented on mobile/ios/WidgetExtension/widgets/RandomWidget.swift:
```diff
@@ -0,0 +1,152 @@
+      provider: ImmichRandomProvider()
+    ) { entry in
+      ImmichWidgetView(entry: entry)
+        .containerBackground(.black, for: .widget)
```
Same as the Memory widget. A dynamic background would be ideal

bwees committed:
fix: dart changes from code review

shenlong-tanwen reviewed (commented):
Thanks a lot for working on this! Most comments are NITs. The overall change looks great

bwees committed:
fix: swift code review changes (not including task groups)

bwees committed:
fix: use task groups to run image retrievals concurrently, get rid of do catch in favor of if let

bwees committed:
chore: cleanup widget service in dart app

bwees committed:
chore: format swift

bwees committed:
fix: remove comma

why does xcode not freak out over this >:(

bwees committed:
switch to preview size for thumbnail

alextran1502 committed:
chore: cropped large image

bwees committed:
fix: properly resize widgets so we dont OOM

bwees committed:
fix: set app group on logout

happens on first install

bwees committed:
fix: stupid app ids

bwees committed:
fix: revert back to thumbnail

we are hitting OOM exceptions due to resizing, once we have on-the-fly resizing on server this can be upgraded

bwees committed:
fix: more memory efficient resizing method, remove extraneous resize commands from API call

bwees committed:
fix: random widget use 12 entries instead of 24 to save memory

bwees committed:
fix: modify duration of entries to 20 minutes and only generate 10 at a time to avoid OOM

bwees committed:
feat: toggle to show album name on random widget

alextran1502 committed:
Podfile lock

Starbix commented:
@bwees How hard would it be to add that when you click on the widget, the app opens on the displayed image?

bwees commented:
> @bwees How hard would it be to add that when you click on the widget, the app opens on the displayed image?

Deep linking will be coming soon in #19232. Once that is in, it will be added!

bwees commented:
@pat 

Make sure your server is updated to the latest version. We only added album filtering on the search API the widget uses in v1.135.0

pat commented:
@bwees thank you that's super helpful, and did indeed get it working! Although now I'm realising I'd prefer to have the widget show a photo from my favourites, which is distinct from albums. 🤔

The roadmap on the website looks relatively high-level - is there a place to log this as a potential (minor) feature? And I guess I could potentially poke around in the typescript code to extend the API… whether any pull request would be of good enough quality is another matter! 😅

bwees commented:
I will put this on my internal todo list. You are more than welcome to take a look and see if you can add the feature yourself!

I believe it should be as simple as extending the Search filters object in the widget to include a filter for favorited photos and enabling that if the user selects a virtual "favorites" album (see the "none" album).
